src/main/sensors/compass.c

```diff
@@ -375,23 +391,34 @@ bool compassIsCalibrationComplete(void)
 
 uint32_t compassUpdate(timeUs_t currentTimeUs)
 {
-    static uint8_t busyCount = 0;
-    uint32_t nextPeriod = COMPASS_READ_US;
-
-    if (busBusy(&magDev.dev, NULL) || !magDev.read(&magDev, magADCRaw)) {
-        // No action was taken as the read has not completed
-        schedulerIgnoreTaskStateTime();
-
-        busyCount++;
-
-        return nextPeriod; // Wait COMPASS_READ_US between states
+    bool checkBusBusy = busBusy(&magDev.dev, NULL);
+    bool checkReadState = !magDev.read(&magDev, magADCRaw);
+
+    // temporary debug for task intervals, only while resolving mag timing issues
+    DEBUG_SET(DEBUG_MAG_TASK_RATE, 4, checkBusBusy);
+    DEBUG_SET(DEBUG_MAG_TASK_RATE, 5, checkReadState);
+
+    static timeUs_t previousTaskTimeUs = 0;
+    const timeDelta_t dTaskTimeUs = cmpTimeUs(currentTimeUs, previousTaskTimeUs);
+    previousTaskTimeUs = currentTimeUs;
+    DEBUG_SET(DEBUG_MAG_TASK_RATE, 6, dTaskTimeUs);
+
+    if (checkBusBusy) {
+        // No action is taken, as the bus was busy.
+        schedulerIgnoreTaskExecRate();
+        return COMPASS_BUS_BUSY_INTERVAL_US; // return delay if the bus is busy
     }
+    if (checkReadState) {
+        // No action is taken, as the compass reported no new data.
+        schedulerIgnoreTaskExecRate();
+        return COMPASS_RECHECK_INTERVAL_US; // return interval if no data is available
+    }
+
+    // if we get here, we have new data
+    // If debug_mode is DEBUG_GPS_RESCUE_HEADING, we should update the magYaw value, after which isNewMagADCFlag will be set false
```

Comment:
(ugly module interdependencies)

Reply:
True; it's a hack.  We only currently use `magYaw` for debugging purposes.  I expect this will be looked into more closely as we move to position hold functionality.